File: python_leetcode/64.py
from typing import List
class Solution:
    # 动态规划方式
    def minPathSum(self, grid: List[List[int]]) -> int:
        m , n  = len(grid),len(grid[0])
        l = m*n
        arr = [0]*l
        arr[0] = grid[0][0]
        for a in range(m):
            for b in range(n):
                if a>0 and b >0 :
                  arr[a*n+b]=grid[a][b]+min(arr[(a-1)*n+b],arr[a*n+b-1])
                elif a ==0 and b>0 :
                    arr[a*n+b]=grid[a][b]+arr[a*n+b-1]
                elif b==0 and a>0:
                    arr[a*n+b]=grid[a][b]+arr[(a-1)*n+b]
        return arr[l-1]
s = Solution()
print(s.minPathSum([[1,2,3],[4,5,6]]))

File: leetcode_c_plus_plus/64.cpp
// 64. 最小路径和
// https://leetcode-cn.com/problems/minimum-path-sum/
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 64.cpp
@Time : 2022/04/03 16:16:57
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
**/
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int row = grid.size(), col = grid[0].size();
        for(int i = 0; i < row; i++) {
            for(int j = 0; j < col; j++) {
                if(i > 0 && j > 0) {
                    grid[i][j] += min(grid[i - 1][j], grid[i][j - 1]);
                } else if(i == 0 && j > 0) {
                    grid[i][j] += grid[i][j - 1];
                } else if(j == 0 && i > 0) {
                    grid[i][j] += grid[i - 1][j];
                }
            }
        }
        return grid[row - 1][col - 1];

    }
};